File: coding-interview-graph-algorithms-enhanced/graph.js
```javascript
class Graph {
  constructor(directed = false) {
    this.adjList = {};
    this.directed = directed;
  }

  addVertex(v) {
    this.adjList[v] = [];
  }

  addEdge(u, v, weight = 1) {
    this.adjList[u].push({ vertex: v, weight });
    if (!this.directed) {
      this.adjList[v].push({ vertex: u, weight });
    }
  }


  //BFS implementation
  bfs(startNode) {
      const queue = [startNode];
      const visited = new Set();
      const distances = {};
      distances[startNode] = 0;
      visited.add(startNode);

      while(queue.length > 0){
          const currentNode = queue.shift();
          for(const neighbor of this.adjList[currentNode]){
              if(!visited.has(neighbor.vertex)){
                  visited.add(neighbor.vertex);
                  queue.push(neighbor.vertex);
                  distances[neighbor.vertex] = distances[currentNode] + 1;
              }
          }
      }
      return distances;
  }

  //DFS implementation -  simplified for brevity.  Add topological sort and cycle detection later.
  dfs(startNode, visited = new Set()){
    visited.add(startNode);
    console.log(startNode); //Process node here.
    for(const neighbor of this.adjList[startNode]){
        if(!visited.has(neighbor.vertex)){
            this.dfs(neighbor.vertex, visited);
        }
    }
  }

  // Dijkstra's algorithm (simplified)
  dijkstra(startNode) {
    const distances = {};
    const visited = new Set();
    for (const node in this.adjList) {
      distances[node] = Infinity;
    }
    distances[startNode] = 0;

    //Implementation needs a priority queue for efficiency
    let minNode = this.findMinDistance(distances, visited);
    while (minNode !== null) {
      for (const neighbor of this.adjList[minNode]) {
        const newDist = distances[minNode] + neighbor.weight;
        if (newDist < distances[neighbor.vertex]) {
          distances[neighbor.vertex] = newDist;
        }
      }
      visited.add(minNode);
      minNode = this.findMinDistance(distances, visited);
    }
    return distances;
  }

  findMinDistance(distances, visited){
    let minDist = Infinity;
    let minNode = null;
    for(const node in distances){
        if(distances[node] < minDist && !visited.has(node)){
            minDist = distances[node];
            minNode = node;
        }
    }
    return minNode;
  }

}


module.exports = Graph;
```

File: algorithm-graph-algorithms-implementation-complete/graph.cpp
#include "graph.h"

template <typename T>
Graph<T>::Graph(int numNodes) : numNodes(numNodes), adjList(numNodes), visited(numNodes, false), recursionStack(numNodes, false) {}

template <typename T>
void Graph<T>::addEdge(int u, int v, int weight) {
    adjList[u].push_back({v, weight});
}


template <typename T>
vector<int> Graph<T>::bfs(int startNode) {
    vector<int> distances(numNodes, -1);
    queue<int> q;
    q.push(startNode);
    distances[startNode] = 0;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (auto& edge : adjList[u]) {
            int v = edge.first;
            if (distances[v] == -1) {
                distances[v] = distances[u] + 1;
                q.push(v);
            }
        }
    }
    return distances;
}


template <typename T>
vector<int> Graph<T>::dfs(int startNode) {
    vector<int> visitedNodes;
    vector<bool> visited(numNodes, false);
    function<void(int)> dfsRecursive = [&](int u) {
        visited[u] = true;
        visitedNodes.push_back(u);
        for (auto& edge : adjList[u]) {
            int v = edge.first;
            if (!visited[v]) {
                dfsRecursive(v);
            }
        }
    };
    dfsRecursive(startNode);
    return visitedNodes;

}


template <typename T>
vector<int> Graph<T>::dijkstra(int startNode) {
    vector<int> distances(numNodes, numeric_limits<int>::max());
    distances[startNode] = 0;
    vector<bool> visited(numNodes, false);

    for (int count = 0; count < numNodes - 1; count++) {
        int u = minDistance(distances, visited);
        visited[u] = true;

        for (auto& edge : adjList[u]) {
            int v = edge.first;
            int weight = edge.second;
            if (!visited[v] && distances[u] != numeric_limits<int>::max() && distances[u] + weight < distances[v]) {
                distances[v] = distances[u] + weight;
            }
        }
    }
    return distances;
}

template <typename T>
int Graph<T>::minDistance(const vector<int>& dist, const vector<bool>& visited) {
    int min = numeric_limits<int>::max(), min_index = -1;
    for (int v = 0; v < numNodes; v++) {
        if (!visited[v] && dist[v] <= min) {
            min = dist[v];
            min_index = v;
        }
    }
    return min_index;
}

//Placeholder for Topological Sort and Cycle Detection (Kahn's Algorithm and DFS based cycle detection would go here)
template <typename T>
vector<int> Graph<T>::topologicalSort() { return {}; } // Placeholder

template <typename T>
bool Graph<T>::isCyclic() { return false; } //Placeholder